### PR TITLE
chore(deps): update bigdecimal from 0.4.1 to 0.4.6

### LINKED ISSUE #10001: [TESTS] Fix sqllogictests to support bigdecimal 0.4.3 crate
### Describe the bug

Since bigdecimal 0.4.3 introduced, there is an display formatting bug leading tests to fail with data  type R.

### To Reproduce

```
query R
select 10.0
----
10

External error: query result mismatch:
[SQL] select 10.0
[Diff] (-expected|+actual)
-   1
+   1E+1
```

### Expected behavior

Test should pass on `bigdecimal` version 0.4.3. 
It works on current `bigdecimal` version 0.4.1

### Additional context

Related https://github.com/apache/arrow-datafusion/pull/9476

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?

Closes #10001.

## Rationale for this change

We use the bigdecimal crate to format decimal values to strings and display them in sqllogical tests(SLTs).
The new version of bigdecimal formats certain decimals in scientific notation, which will break many SLT tests.

There are two possible solutions:
1. Set the compile-time environment variables for bigdecimal, such as `RUST_BIGDECIMAL_FMT_EXPONENTIAL_LOWER_THRESHOLD`. However, I'm a bit concerned about whether it's appropriate to operate environment variables in a library like DataFusion. Additionally, we might need to introduce a separate Cargo config.toml into the project root directory so that `cargo test --test sqllogictests` can work.
2. Implement our own method for formatting decimals.

This PR takes the latter approach.
## Are these changes tested?
Yes. By new UTs and existing SLT tests.

## Are there any user-facing changes?
No. Only modified the test engine.

### CODE REVIEW

**Review by comphead — APPROVED** [review]
thanks @jonahgao for taking care on this

**Review by Weijun-H — APPROVED** [review]
LGTM! Thanks @jonahgao, it is an elegant solution 👍

**Comment by jonahgao** [comment]
Thanks @comphead @Weijun-H

**Comment by comphead** [comment]
Btw, I was experimenting with env var, I didn't work well

**Comment by jonahgao** [comment]
> Btw, I was experimenting with env var, I didn't work well

I tested it on this [branch](https://github.com/jonahgao/datafusion/commit/92345ea07a4a964a2465b234addac699a035e647#diff-9a4f3e4537ebb7474452d131b0d969d89a51286f4269aac5ef268e712be17268R2), and it works except for one query.
`cargo clean -p bigdecimal` might be needed due to https://github.com/rust-lang/cargo/issues/10358 (bigdecimal reads envs in build.rs)

**Comment by comphead** [comment]
Good catch

File: datafusion/sqllogictest/src/engines/conversion.rs [hunk 1]
```diff
@@ -101,5 +101,70 @@ pub(crate) fn decimal_to_str(value: Decimal) -> String {
 }
 
 pub(crate) fn big_decimal_to_str(value: BigDecimal) -> String {
-    value.round(12).normalized().to_string()
+    // Round the value to limit the number of decimal places
+    let value = value.round(12).normalized();
+    // Format the value to a string
+    format_big_decimal(value)
```
Comment (findepi) [inline datafusion/sqllogictest/src/engines/conversion.rs:107]: we could use bigdecimal `to_plain_string` (see https://github.com/akubera/bigdecimal-rs/issues/135) -- but it's not released yet afaict

**Pull request merged**